homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,224 @@
+            idx=idx,
+            coordinator=ws66i_data.coordinator,
+        )
+        for idx, zone_id in enumerate(ws66i_data.zones)
+    )
+
+    # Set up services
+    platform = async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SNAPSHOT,
+        {},
+        "snapshot",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_RESTORE,
+        {},
+        "async_restore",
+    )
+
+
+class Ws66iZone(CoordinatorEntity, MediaPlayerEntity):
+    """Representation of a WS66i amplifier zone."""
+
+    def __init__(
+        self,
+        device: WS66i,
+        ws66i_data: Ws66iData,
+        entry_id: str,
+        zone_id: int,
+        idx: int,
+        coordinator: Ws66iDataUpdateCoordinator,
+    ) -> None:
+        """Initialize a zone entity."""
+        super().__init__(coordinator)
+        self._ws66i: WS66i = device
+        self._ws66i_data: Ws66iData = ws66i_data
+        self._zone_id: int = zone_id
+        self._zone_id_idx: int = idx
+        self._coordinator = coordinator
+        self._snapshot: ZoneStatus = None
+        self._status: ZoneStatus = coordinator.data[idx]
+        self._attr_source_list = ws66i_data.sources.name_list
+        self._attr_unique_id = f"{entry_id}_{self._zone_id}"
+        self._attr_name = f"Zone {self._zone_id}"
+        self._attr_supported_features = SUPPORT_WS66I
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, str(self.unique_id))},
+            name=self.name,
+            manufacturer="Soundavo",
+            model="WS66i 6-Zone Amplifier",
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        # This will be called for each of my entities after the coordinator
+        # finishes executing _async_update_data()
+
+        # Save a reference to the zone status that this entity represents
+        self._status = self.coordinator.data[self._zone_id_idx]
+
+        # Parent will notify HA that our fields were updated
+        super()._handle_coordinator_update()
+
+    @property
+    def state(self) -> str:
+        """State of the player."""
+        return STATE_ON if self._status.power else STATE_OFF
+
+    @property
+    def volume_level(self) -> float:
+        """Volume level of the media player (0..1)."""
+        return self._status.volume / float(MAX_VAL)
+
+    @property
+    def is_volume_muted(self) -> bool:
+        """Boolean if volume is currently muted."""
+        return self._status.mute
+
+    @property
+    def media_title(self) -> str:
+        """Title of current playing media."""
+        return self._ws66i_data.sources.id_name[self._status.source]
+
+    @property
+    def source(self) -> str:
+        """Name of the current input source."""
+        # Source name is found from the dataclass
+        return self._ws66i_data.sources.id_name[self._status.source]
```

Comment:
You could drop these and set them as `self._attr_X` in `_handle_coordinator_update` instead

Reply:
Example https://github.com/bdraco/home-assistant/commit/e777ed57ef23b7aef9dffb21732df0dc522201d2

Reply:
I like it.